Integrate CUJ 3 documentation into components.md (#3350)
kubeflow/website: content/en/docs/components/pipelines/author-a-pipeline/components.md

**When to use?** Containerized Python components should be used any time your component is implemented as Python code, but cannot be written as a standalone Python function or you wish to organize source code outside of the component Python function definition.

### 3. Custom container components

> Container-based components make it easier to use the pre-existed container to build a KFP component. 

Reviewer: Consistency of language: `Custom container components`

> The user will control the container entrypoint and is able to define the component by specifying their container, commands, and args in a `ContainerSpec` object with similar authoring experiences as defining Python components. 

Reviewer: This document addresses the reader directly, saying "you".

Can you update your doc to take this style? E.g., `You can control the container entrypoint`

> The user will control the container entrypoint and is able to define the component by specifying their container, commands, and args in a `ContainerSpec` object with similar authoring experiences as defining Python components. 

Reviewer: "image, command, and args"

To define a container component, you must:

> 1) Write your component’s code as a python function that returns a `dsl.ContainerSpec` object to specify the container image and the commands to be run in the container and wrap the function into a `@container_component` decorator. The function should satisfy the following requirements:

Reviewer: `python` -> `Python`

>     *   It should not use any code declared outside of the function definition.

Reviewer: nit:
This is not strictly true, as it uses `dsl.ContainerSpec` which is declared outside of the function definition. Also, the next bullet about `dsl.ContainerSpec` conveys this idea: "the function should do nothing other than..."

WDYT about removing the first bullet and turning the second bullet into a paragraph?

Definitely a nit, so feel free to proceed as-is if you prefer it that way

        object, with the following parameters:
        * image: The target image where the command (with args) is executed. The user will control the entrypoint of the image.
        * command(optional): The command to be executed. 
        * args(optional): The arguments of the command. It’s recommended to place the input of the     components in the args section instead of the command section.

    The decorator will then compose a `ComponentSpec` object using the `ContainerSpec` to be compiled for execution. (Learn more about [ContainerSpec][dsl-reference-documentation] in documentation.) 

2) Specify your function's inputs and outputs in the function's    signature [Learn more about passing data between components][data-passing]. Specifically for container-based component, your function's inputs and outputs must meet the following requirements:
    
    *   All your function's arguments must have data type annotations.
    *   Different from a Python component, your return type annotation for the function 
        must either be `dsl.ContainerSpec` or omitted. 
    *   If the function accepts or returns large amounts of data or complex

>         data types, you must annotate that argument as an _artifact_. Note that in the function you defined, you can only access artifacts via its `.url`, `.path`, or `.metadata` attribute. Accessing any other attribute or the artifact variable by itself is currently not allowed. 

Reviewer: nit: remove "currently" of "currently not allowed"

"currently" suggests we may soon add this feature, which we do not have plans to do.

Here is a minimal example that uses a bash script to demonstrate the use of container component:
```python
from kfp import dsl
@dsl.container_component
def hello_world_comp(text: str, out_filename: dsl.OutputPath(str)):
    return dsl.ContainerSpec(
        image='python:3.7' # or put your custom container image here
        command=['source', 'hello_world.sh'],
        args=['--in', text, '--out', out_filename]
    )
```

> Meanwhile, add a bash script `hello_world.sh` on your container that takes outputs the text into the provided path:

Reviewer: This is a tricky example, since `python:3.7` is a pre-existing image that will not include `hello_world.sh` and there is no way to inject it into the container. Additionally, I think this example would fail, because the `$4` file does not exist until you create it.

What do you think about removing this example, then using the pipeline below with a bit more detail on each component?

Author: I guess it's actually okay to use `>>` for piping to create a new file if the file doesn't exist yet. But I agree that the pipeline below would be more suitable in this case.

```bash
echo $2 >> $4
```
Below is a more complex example that authors a pipelines from two container component. Just as using with a Python component, we can access the outputs of a `container_component` for downstream tasks as demonstrated in the pipeline:
```python
from kfp.dsl import (
  container_component,
  ContainerSpec,
  Dataset,
  Input, 
  Output,
) 

@container_component
def component1(text: str, output_gcs: Output[Dataset]):
    return ContainerSpec(
        image='alpine',
        command=[
            'sh',
            '-c',
            'mkdir --parents $(dirname "$1") && echo "$0" > "$1"',
        ],
        args=[text, output_gcs.path])

@container_component
def component2(input_gcs: Input[Dataset]):
    return ContainerSpec(image='alpine', command=['cat'], args=[input_gcs.path])

> @pipeline(name='two-step-pipeline-containerized')

Reviewer: `pipeline` is not defined

> @pipeline(name='two-step-pipeline-containerized')

Reviewer: nit: maybe remove `name` argument to be consistent with the other examples in this doc -- just trying to minimize examples to the bare necessities to improve clarity

### 3. Custom container components

> Custom container components make it easier to use the pre-existed container to build a KFP component. 

Reviewer: Can you describe what you mean by "the pre-existed container"? I think a slightly more accurate description would be "Custom container components allow you to specify a container to execute as your component." or something similar. WDYT?

Author: Yes, I agree that's better phrasing! I will correct that.

> You can control the container entrypoint and define the custom container component by specifying their image, commands, and args in a `ContainerSpec` object with similar authoring experiences as defining Python components. 

Reviewer: What do you think about simplifying this tot he following:

```suggestion
The `dsl.ContainerSpec` object allows you to specify a container via an image, command, and args.
```

To define a container component, you must:
1) Write your component’s code as a Python function that returns a `dsl.ContainerSpec` object to specify the container image and the commands to be run in the container and wrap the function into a `@container_component` decorator. The function should do nothing other than returning a `dsl.ContainerSpec` object, with the following parameters:

>     * image: The target image where the command (with args) is executed. You will control the entrypoint of the image.

Reviewer: nit: enclose parameter names in backticks to stylize as code

>     * image: The target image where the command (with args) is executed. You will control the entrypoint of the image.

Reviewer: I suggest a slight rephrasing. The command is executed in the _container_, which runs an image. And you _may_ control the entrypoint, but do not need to (optional args).

```suggestion
    * image: The image that the container will run. You can use `command` and `args` to control the entrypoint.
```

    * command(optional): The command to be executed. 
    * args(optional): The arguments of the command. It’s recommended to place the input of the components in the args section instead of the command section.

>     The decorator will then compose a `ComponentSpec` object using the `ContainerSpec` to be compiled for execution. (Learn more about [ContainerSpec][dsl-reference-documentation] in documentation.) 

Reviewer: We haven't spoken much about a `ComponentSpec` object so far and it's not part of our public API. WDYT about just explaining that this component can be used like the others, omitting any reference to a `ComponentSpec`?

def component2(input_gcs: Input[Dataset]):
    return ContainerSpec(image='alpine', command=['cat'], args=[input_gcs.path])

@pipeline
def two_step_pipeline_containerized(text: str):
    component_1 = component1(text)
    component_2 = component2(input_gcs=component_1.outputs['output_gcs'])
```

> In the above example, `component1` takes in a text and output it to a path as an artifact. Then, the `component2` retrieves the artifact output by the `component1` and prints it out.

Reviewer: nit: WDYT about renaming these so that they are more descriptive of what they do? Maybe something like `create_dataset` and `print_dataset`?
